lib/public_key/src/pubkey_policy_tree.erl

```diff
@@ -0,0 +1,421 @@
+                           end, [], SubTree);
+valid_policy_node_set({#{valid_policy := ?anyPolicy}, Leaves}) when is_list(Leaves) ->
+    Leaves;
+valid_policy_node_set(_) ->
+    [].
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+any_node() ->
+    policy_node(?anyPolicy, [], [?anyPolicy]).
+
+any_leaves({_,[{_, _}|_] = SubTree}) ->
+    lists:flatmap(fun(ChildNode)->
+                          any_leaves(ChildNode)
+                  end, SubTree);
+any_leaves({_, Leaves}) ->
+    AnyLeaf = fun(#{valid_policy := ?anyPolicy} = Node) ->
+                      {true, Node};
+                 (_) ->
+                      false
+              end,
+    lists:filtermap(AnyLeaf, Leaves).
+
+any_leaf(Tree) ->
+    case any_leaves(Tree) of
+        [] ->
+            no_node;
+        AnyLeaves ->
+            AnyLeaves
+    end.
+
+collect_subtree_qualifiers(_, [], _) ->
+    [];
+collect_subtree_qualifiers(Collect, [{#{expected_policy_set := Set}, _} = ChildNode | SubTree], Policy) ->
+    case lists:member(Policy, Set) of
+        true ->
+            lists:flatten(subtree_collect(Collect, ChildNode));
+        false ->
+            collect_subtree_qualifiers(Collect, SubTree, Policy)
+    end;
+collect_subtree_qualifiers(Collect, SubTree, _) ->
+    lists:flatten(subtree_collect(Collect, SubTree)).
+
+subtree_collect(Collect, {Parent, SubTree})  ->
+    Collect(Parent) ++ subtree_collect(Collect, SubTree);
+subtree_collect(Collect, [ {_,_} = ChildNode | SubTree]) ->
+    subtree_collect(Collect, ChildNode) ++ subtree_collect(Collect, SubTree);
+subtree_collect(Collect, Leaves) ->
+    lists:map(Collect, Leaves).
+
+constrain({#{valid_policy := ?anyPolicy}, [{_, _}| _] = SubTree})  ->
+    Parents = [Parent || {Parent, _} <- SubTree],
+    lists:filtermap(fun(#{valid_policy := ?anyPolicy}) ->
+                            false;
+                       (Node) ->
+                            {true, Node}
+                    end, Parents) ++
+        lists:foldl(fun(Child, Acc) ->
+                            constrain(Child) ++ Acc
+                    end, [], SubTree);
+constrain({#{valid_policy := ?anyPolicy}, Leaves}) when is_list(Leaves) ->
+    lists:filtermap(fun(#{valid_policy := ?anyPolicy}) ->
+                            false;
+                       (Node) ->
+                            {true, Node}
+                    end, Leaves);
+constrain(_) ->
+    [].
+
+filter_leaves({Parent,[{_, _}|_] = SubTree}, LeafFun) ->
+    {Parent, lists:map(fun(ChildNode)->
+                               filter_leaves(ChildNode, LeafFun)
+                       end, SubTree)};
+filter_leaves({Parent, Leaves}, LeafFun) ->
+    {Parent, lists:filtermap(LeafFun, Leaves)}.
+
+depth({_, SubTree}) ->
+    depth(SubTree, 0).
+
+depth([], N)-> 
+    N;
+depth([_| Rest], N) ->
+    depth(Rest, N+1).
```

Comment:
this does horizontal traversal of nodes children. AFAIK it calculates width of a subtree at 2nd level.
depth is not calculated here.

Reply:
Maybe I had a little to many thing to do today!